### ISSUE TITLE
Hasura Allow list plugin: support fragments defined globally

### ISSUE THREAD

**Issue by BenoitRanque** [dotansimha/graphql-code-generator#7700]
**Is your feature request related to a problem? Please describe.**

Currently the hasura allow list plugin requires that fragments be defined in the same document as the operations that use them.

**Describe the solution you'd like**

It would be better if instead, fragments were pulled from all documents. This works as long as fragment names are unique throughout the project.

**Comment by shinnoki** [comment]
+1
We use hasura-allow-list plugin but patch it by ourself for this issue.

Note that Hasura's allow lists strictly respect the order of documents, fragments and fields.
I think it would be better if adopt Visitor Pattern, because this plugin is usually used together with other typescript plugins such as typed-document-node, and it's better to be guaranteed mechanically to behave the same way as them.

**Comment by franciscolourenco** [comment]
> This works as long as fragment names are unique throughout the project.

@BenoitRanque to clarify, are you saying that the plugin currently support fragments defined in separate files, as long as each fragment has a unique name?

**Comment by BenoitRanque** [comment]
> > This works as long as fragment names are unique throughout the project.
> 
> @BenoitRanque to clarify, are you saying that the plugin currently support fragments defined in separate files, as long as each fragment has a unique name?
> 

Yes, if the global fragments option is enabled, the plug-in will recursively append fragments found in processed files (usually a target directory) to the generated allow list entry.

This requires that that fragment name be unique across your project as that is how required fragments are identified.
